aws_signing_helper/update.go

```diff
@@ -98,90 +98,92 @@ func GetCredentialsFileContents() ([]string, error) {
+			writeLines = append(writeLines[:], profileSection+"\n")
+			readLinesIndex += 1
+			for ; readLinesIndex < len(readLines); readLinesIndex++ {
 				// If the last line of the credentials file is reached
 				// OR the next profile section is reached
-				if j == len(writeLines)-1 || strings.HasPrefix(writeLines[j], "[") {
+				if readLinesIndex == len(readLines)-1 || strings.HasPrefix(readLines[readLinesIndex], "[") {
 					if !newCredVisit["aws_access_key_id"] {
-						_, err = destFileWriter.WriteString(accessKey)
+						writeLines = append(writeLines[:], accessKey)
 					}
 					if !newCredVisit["aws_secret_access_key"] {
-						_, err = destFileWriter.WriteString(secretKey)
+						writeLines = append(writeLines[:], secretKey)
 					}
 					if !newCredVisit["aws_session_token"] {
-						_, err = destFileWriter.WriteString(sessionToken)
+						writeLines = append(writeLines[:], sessionToken)
 
 					}
-					if j == len(writeLines)-1 {
-						i = j
-					} else {
-						i = j - 1
+					if readLinesIndex != len(readLines)-1 {
+						readLinesIndex -= 1
 					}
 					profileExist = true
 					break
-				} else if strings.HasPrefix(writeLines[j], "aws_access_key_id") {
+				} else if strings.HasPrefix(readLines[readLinesIndex], "aws_access_key_id") {
 					// replace "aws_access_key_id"
-					_, err = destFileWriter.WriteString(accessKey)
+					writeLines = append(writeLines[:], accessKey)
 					newCredVisit["aws_access_key_id"] = true
-				} else if strings.HasPrefix(writeLines[j], "aws_secret_access_key") {
+				} else if strings.HasPrefix(readLines[readLinesIndex], "aws_secret_access_key") {
 					// replace "aws_secret_access_key"
-					_, err = destFileWriter.WriteString(secretKey)
+					writeLines = append(writeLines[:], secretKey)
 					newCredVisit["aws_secret_access_key"] = true
-				} else if strings.HasPrefix(writeLines[j], "aws_session_token") {
+				} else if strings.HasPrefix(readLines[readLinesIndex], "aws_session_token") {
 					// replace "aws_session_token"
-					_, err = destFileWriter.WriteString(sessionToken)
+					writeLines = append(writeLines[:], sessionToken)
 					newCredVisit["aws_session_token"] = true
 				} else {
 					// write other keys
-					_, err = destFileWriter.WriteString(writeLines[j] + "\n")
+					writeLines = append(writeLines[:], readLines[readLinesIndex]+"\n")
 				}
 			}
 		} else {
-			_, err = destFileWriter.WriteString(writeLines[i] + "\n")
+			writeLines = append(writeLines[:], readLines[readLinesIndex]+"\n")
 		}
 	}
 
 	// If the chosen profile does not exist
 	if !profileExist {
 		writeCredential := profileSection + "\n" + accessKey + secretKey + sessionToken
-		_, err = destFileWriter.WriteString(writeCredential + "\n")
+		writeLines = append(writeLines[:], writeCredential+"\n")
 	}
 
-	if err != nil {
-		return err
+	// Create buffered writer with a buffer of maximum size
+	destFileWriter := bufio.NewWriterSize(destFile, math.MaxInt32)
```

Comment:
Why max size?

Reply:
I thought it would be better this way since the write would then effectively be all or nothing. But maybe that isn't correct.

Reply:
The docs say it creates a buffer of at least that size; MaxInt32 is ~ 2GB, which is will tax less powerful machines. The buffer does not limit the size of the file, only how much is held in memory before flushing to disk. A few dozen KB is probably fine.